intro to command line (#9)
Ada-Developers-Academy/core: intro-to-dev-environment/intro-to-command-line.md

- Apply the python3 command in the command line to run a Python script.

## Introduction

As a programmer, the terminal is one of our most important tools, and we will run thousands of commands in the terminal over our career. Before we get there, we should spend time familiarizing ourselves with the command line.

## Vocabulary and Synonyms

| Vocab          | Definition                                                                                                                                                     | Synonyms  | How to Use in a Sentence                                                                                                                                                                                                         |
| -------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | interface      | A medium that takes information from one source, and passes it to another source.The interface usually makes a meaningful translation between the two sources. | -         | "The command line and the terminal is one interface for how we interact with the file system and run programs on our computer," "That program has a graphical user interface, and there are buttons and images to interact with" |

Reviewer: ```suggestion
| interface      | A medium that takes information from one source, and passes it to another source. The interface usually makes a meaningful translation between the two sources. | -         | "The terminal is one interface for how we interact with the file system and run programs on our computer", "That program has a graphical user interface, and there are buttons and images to interact with" |
```

Author: The space is a good catch!

In terms of English grammar, I feel fairly confident that the commas should go inside of the quotation marks ([grammarly rules](https://www.grammarly.com/blog/quotation-marks) under the header "quotation marks and other punctuation marks")

I could see us putting the comma on the outside of the quote if we're thinking that the quote is a string literal ;) lolol, then it might make sense to me if the comma is outside. Or do you have another reason?

Reviewer: LOL Yes! I was thinking of this column containing 'lists' of sentences.

> | folder         | A named container of files and other folders (normally named sub-folders). Can be inside another folder.                                                       | directory | "I'm going to make another folder in this folder to organize things better," "Check to see if you put that file in the home directory"                                                                                           |

Reviewer: ```suggestion
| folder         | A named container of files and other folders (normally named sub-folders). Can be inside another folder.                                                       | directory | "I'm going to make another folder in this folder to organize things better", "Check to see if you put that file in the home directory"                                                                                           |
```

> | Root directory | The nickname of the top-most folder that exists on the machine                                                                                                 | root      | "That program gets installed and its files live somewhere in root," "I don't usually spend much time in the root directory"                                                                                                      |

Reviewer: ```suggestion
| Root directory | The nickname of the top-most folder that exists on the machine                                                                                                 | root      | "That program gets installed and its files live somewhere in root", "I don't usually spend much time in the root directory"                                                                                                      |
```

## Recommended File System Structure

As developers, over time, we will make hundreds of files and folders.

If you've never organized that many files and folders, and don't have a preference for how to organize, then we recommend that following this folder structure:

### `~/Developer`

- This is a folder where you put all of your developer-y code stuff!
- Make this folder with `$ mkdir ~/Developer` for the first time
- Navigate to this folder from any location in Terminal with `$ cd ~/Developer`, or using your relative commands to change directories

### `~/Developer/projects`

- This is a folder where you put all of your Ada projects
- Make this folder with `$ mkdir ~/Developer/projects` for the first time
- Navigate from anywhere with `$ cd ~/Developer/projects`

### `~/Developer/classroom`

- This is a folder where you put all of your notes, exercises, scratchwork, classroom activities, homework, etc
- Make this folder with `$ mkdir ~/Developer/classroom`
- Pick one of these two popular ways to organize your notes:
  - Organize notes by week. They will have sub-folders nested in here, such as `~/Developer/classroom/week-1`

>   - Organize notes by topic. Recommended topic folder names are `git`, `cs-fun`, `ruby`, `rails`, `js`, `react`, so it is like `~/Developer/classroom/ruby`

Reviewer: ```suggestion
  - Organize notes by topic. Recommended topic folder names are `git`, `cs-fun`, `python`, `flask`, `js`, `react`, so it is like `~/Developer/classroom/python`
```

Reviewer: no more ruby 😆

Of course, feel free to not follow these recommendations if you have preferred ways of organizing files that will work for you better.

## Always Run Code with `python3`

This curriculum teaches programming in Python 3. By default, many macOS systems come with Python 2 installed. However, the differences between Python 3 and Python 2 are significant! It will become **crucial** to always execute our Python scripts from the command line and specify Python 3.

We can specify that we want to run specifically Python 3, we will always use this command:

```bash
python3 some_file.py
```

...where `some_file.py` is the (path and) name of the Python script we want to run.

If we forget to run our Python scripts with `python3`, there's a chance that the script runs as expected, and a chance that the script breaks or has bugs.

> Callout: "A lot of tutorials just use the `python` command to run Python scripts! Why? I'm not convinced that we need to use the `python3` command." One way to check if you need to use the `python3` command is to run two commands: `$ python --version` and `$ python3 --version`. If the output of these two commands differs (namely, `python` points to version 2), then we need to **avoid** using the `python` command.
> 
> Callout: There are many methods to make shortcuts for the `python3` command, including terminal/bash aliases, and more. Follow your curiosity!

Reviewer: ```suggestion
### !callout-info

## python3
"A lot of tutorials just use the `python` command to run Python scripts! Why? I'm not convinced that we need to use the `python3` command." One way to check if you need to use the `python3` command is to run two commands: `$ python --version` and `$ python3 --version`. If the output of these two commands differs (namely, `python` points to version 2), then we need to **avoid** using the `python` command.

### !end-callout

### !callout-info

## python3 
There are many methods to make shortcuts for the `python3` command, including terminal/bash aliases, and more. Follow your curiosity!

### !end-callout
```

## Continue Learning

It is crucial for our careers as developers to practice using the terminal, and to increase our skill in it. There are hundreds of more commands to learn.

We can make a commitment to ourselves to discover and learn more and more ways to use the command line, to help our development workflow become more and more robust.

Some commands to learn include:

- `man`
- `mv`
- `cat`
- `echo`
- `grep`

## Exiting Weird Screens on Terminal

Terminal and some commands bring us to different text editors. If you're ever taken to a terminal screen where:

- arrow keys or key presses don't seem to work as expected
- there is a `:` on the bottom line
- `ctrl+c` does not exit this screen

> Try using the up and down arrows to scroll, and tapping the key `q` to quit.

Reviewer: Is this referring to Vim? If so, we might need to explain Vim a LITTLE bit.

Author: ```suggestion
Try using the up and down arrows to scroll, and tapping the key `q` to quit. If you're ever in this situation, macOS's default text editor, `vim` is at play here!
```

What do you think about this? Deeper?

Reviewer: ```suggestion
Try using the up and down arrows to scroll, and tapping the `q` or `esc` key to quit.  If you're ever in this situation, one of macOS's default text editors, `[vim](https://en.wikipedia.org/wiki/Vim_(text_editor)` is at play here!

See more vim commands [here](https://devhints.io/vim).
```

Something like that. Nano(ew) is another text editor that comes with MacOS but I'm not sure if that fact is important for the context. https://support.apple.com/guide/terminal/use-command-line-text-editors-apdb02f1133-25af-4c65-8976-159609f99817/mac.

Author: ```suggestion
Try using the up and down arrows to scroll, and tapping the `q` or `esc` key to quit.  If you're ever in this situation, one of macOS's default text editors, [`vim`](https://en.wikipedia.org/wiki/Vim_(text_editor) is at play here! See more vim commands [with this vim resource](https://devhints.io/vim).
```

Adjusted this for two details @audreyandoy :

1. Your syntax for the vim link wasn't quite correct-- the backticks should be around the link text!
2. It's a preference to not have a link with the text "here" as it isn't quite descriptive (think web accessibility).

Check out my suggestion and lmk if you have questions on those style preferences!

Reviewer: oh!! Thank you for reminding me about web accessibility! That suggestion looks great. :)

> ## Check for Understanding

Reviewer: ```suggestion
## Check for Understanding

## Check for Understanding
<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->

### !challenge

* type: multiple-choice
* id: 0d18747d-94a0-4136-85b1-49a1308ef647
* title: The command line 
<!-- * points: [1] (optional, the number of points for scoring as a checkpoint) -->
<!-- * topics: [python, pandas] (optional the topics for analyzing points) -->

##### !question

What is the command line? 

##### !end-question

##### !options

* The interface where developers type in commands into a terminal
* The top-most folder that exists on a machine
* a program that runs commands

##### !end-options

##### !answer

* The interface where developers type in commands into a terminal

##### !end-answer

### !end-challenge

<!-- ======================= END CHALLENGE ======================= -->
```

Reviewer: I'll probably add more if necessary
